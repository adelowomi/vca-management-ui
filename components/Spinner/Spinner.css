.spinner {
  display: block;
}
.spinner .progress {
  transition: 300ms stroke-dashoffset;
  transform: rotate(-90deg);
  transform-origin: 50% 50%;
  stroke: #0035c1;
}
.spinner .circle {
  stroke: rgba(0, 0, 0, 0.2);
}
.spinner.spinner-indeterminate .spinner-inner {
  transform: translateZ(0);
  animation: load8 1.1s infinite linear;
  -webkit-animation: load8 1.1s infinite linear;
  -ms-transform: translateZ(0);
  -webkit-transform: translateZ(0);
}

@-webkit-keyframes load8 {
  0% {
    transform: rotate(0deg);
    -webkit-transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
  }
}

@keyframes load8 {
  0% {
    transform: rotate(0deg);
    -webkit-transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
  }
}
