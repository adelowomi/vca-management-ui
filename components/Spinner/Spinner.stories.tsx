import { number, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import Spinner from './Spinner';

storiesOf('Spinner', module)
  .addDecorator(withKnobs)
  .add('Spinner Demo', () => (
    <div className="spinners">
      <div>
        Small
        <Spinner size="small" />
      </div>
      <div>
        Medium
        <Spinner size="medium" />
      </div>
      <div>
        Large
        <Spinner size="large" />
      </div>
    </div>
  ))
  .add('Spinner with progress', () => (
    <Spinner
      size="medium"
      progress={number('Progress', 10, { range: true, min: 0, max: 100 })}
    />
  ));
