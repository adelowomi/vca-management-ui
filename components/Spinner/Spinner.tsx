import './Spinner.css';

import React, { useMemo } from 'react';

export const SIZE_ENUM = ['small', 'medium', 'large'];

const sizeMap = {
  small: {
    center: 10,
    dashArray: 56.5,
    radius: 9,
    strokeWidth: 2,
    widthHeight: 20,
  },
  medium: {
    center: 21,
    dashArray: 113,
    radius: 18,
    strokeWidth: 3,
    widthHeight: 42,
  },
  large: {
    center: 42,
    dashArray: 226,
    radius: 36,
    strokeWidth: 6,
    widthHeight: 84,
  },
};

export type SpinnerSizeTypes = 'small' | 'medium' | 'large';

export interface SpinnerProps {
  /**
   *Size of the Spinner
   */
  size?: SpinnerSizeTypes;
  /**
   * Progress is used in a the load progress case.
   * Default is -1 which internally defaults to using a indeterminate spinner.
   */
  progress?: number;
}

const Spinner: React.FC<SpinnerProps> = ({ size, progress }) => {
  const { center, dashArray, radius, strokeWidth, widthHeight } = sizeMap[size];

  const isProgress = Boolean(progress > 0);

  const offset = useMemo(() => {
    const internalProgress = isProgress ? progress : 25;
    return ((100 - internalProgress) / 100) * dashArray;
  }, [dashArray, isProgress, progress]);

  const classNames = useMemo(() => {
    const spinnerClassNames = ['spinner'];
    if (!isProgress) {
      spinnerClassNames.push('spinner-indeterminate');
    }
    return spinnerClassNames.join(' ');
  }, [isProgress]);

  const circleProps = {
    cx: center,
    cy: center,
    r: radius,
    fill: 'transparent',
    strokeDasharray: dashArray,
    strokeWidth,
  };

  return (
    <div className={classNames} data-testid={`${size}-${progress}`}>
      <svg className="spinner-inner" height={widthHeight} width={widthHeight}>
        <circle className="circle" {...circleProps} />
        <circle
          className="progress"
          {...circleProps}
          strokeDashoffset={offset}
        />
      </svg>
    </div>
  );
};

Spinner.defaultProps = {
  size: 'small',
  progress: -1,
};

export default Spinner;
