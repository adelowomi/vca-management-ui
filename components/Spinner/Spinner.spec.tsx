import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import React from 'react';

import Spinner, { SpinnerSizeTypes } from './Spinner';

describe('<Spinner />', () => {
  const sizes: SpinnerSizeTypes[] = ['small', 'medium', 'large'];
  describe.each([sizes])('Given default (%s) size spinner', (size) => {
    const { container } = render(<Spinner size={size} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('Given progress is set', () => {
    const { container, getByTestId } = render(
      <Spinner progress={10} size="small" />
    );
    expect(container.firstChild).toMatchSnapshot();
    it('then progress is correctly represented', () => {
      expect(getByTestId('small-10')).toBeTruthy();
    });
  });

  describe('Given progress is not set', () => {
    it('then progress is correctly represented', () => {
      const { queryByTestId } = render(<Spinner />);
      expect(queryByTestId('small-10')).toBeFalsy();
    });
  });
});
