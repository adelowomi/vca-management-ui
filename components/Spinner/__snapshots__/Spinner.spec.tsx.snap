// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  class="spinner spinner-indeterminate"
  data-testid="small--1"
>
  <svg
    class="spinner-inner"
    height="20"
    width="20"
  >
    <circle
      class="circle"
      cx="10"
      cy="10"
      fill="transparent"
      r="9"
      stroke-dasharray="56.5"
      stroke-width="2"
    />
    <circle
      class="progress"
      cx="10"
      cy="10"
      fill="transparent"
      r="9"
      stroke-dasharray="56.5"
      stroke-dashoffset="42.375"
      stroke-width="2"
    />
  </svg>
</div>
`;

exports[` 2`] = `
<div
  class="spinner"
  data-testid="small-10"
>
  <svg
    class="spinner-inner"
    height="20"
    width="20"
  >
    <circle
      class="circle"
      cx="10"
      cy="10"
      fill="transparent"
      r="9"
      stroke-dasharray="56.5"
      stroke-width="2"
    />
    <circle
      class="progress"
      cx="10"
      cy="10"
      fill="transparent"
      r="9"
      stroke-dasharray="56.5"
      stroke-dashoffset="50.85"
      stroke-width="2"
    />
  </svg>
</div>
`;
